### PR TITLE
devop40: deploy krakens on multiple machines

### PR DESCRIPTION
**Pull request by francois-vincent**

### CODE REVIEW

File: fabfile/component/kraken.py [hunk 1]
```diff
@@ -219,18 +217,18 @@ def purge_data_nav(force=False):
...
     dead = 0
     threshold = env.kraken_threshold * len(env.instances.values())
     for instance in env.instances.values():
-        # env.host will call the monitor kraken on the current host
-        request = 'http://{}:{}/{}/?instance={}'.format(env.host,
-            env.kraken_monitor_port, env.kraken_monitor_location_dir, instance.name)
-        result = _test_kraken(request, fail_if_error=False)
-        if not result or result['status'] == 'timeout' or result['loaded'] is False:
-            dead += 1
+        for host in (e.split('@')[0] for e in instance.kraken_engines):
```
Comment (antoine-de) [inline fabfile/component/kraken.py]: I think it's better to have a separate field in instance, I don't like this `split` in the middle of the code

**Comment by kinnou02** [comment]
It would be nice to have some kind of explanation about what this PR do and how.

File: fabfile/component/kraken.py [hunk 2]
```diff
@@ -317,61 +318,69 @@ def _test_kraken(query, fail_if_error=True):
...
-            result = Retrying(stop_max_delay=env.KRAKEN_RESTART_DELAY * 1000,
-                              wait_fixed=1000, retry_on_result=lambda x: x is None or not x['loaded']) \
-                .call(_test_kraken, request, fail_if_error)
-        except Exception as e:
-            print(red("ERROR: could not reach {}, too many retries ! ({})".format(instance.name, e)))
-            result = {'status': False}
-    else:
-        result = _test_kraken(request, fail_if_error)
+    hosts = hosts or (e.split('@')[1] for e in instance.kraken_engines)
+    will_return = len(hosts) == 1
```
Comment (antoine-de) [inline fabfile/component/kraken.py]: I'm not sure it's very straigh forward, you return a value only if there is one host ?

i'm not fond of this

Comment (francois-vincent) [inline fabfile/component/kraken.py]: I completely agree with this opinion, but that's how this code is working now and I don't wish to mix deep code refactoring within this PR.
This should be done in another specific refactoring PR (with a Jira ticket)
Here is my suggestion:
I have a good workable plan to introduce very good unit tests to this project.
We could open a new ticket to add unit tests to the "kraken" component AND to refactor this component in the same time. With unit tests developed first, refactoring will be quicker and safer.

File: fabfile/component/kraken.py (hunk 1 above)
Comment (francois-vincent) [inline fabfile/component/kraken.py]: right

File: fabfile/component/kraken.py (hunk 2 above)
Comment (francois-vincent) [inline fabfile/component/kraken.py]: @kinnou02 : There's a ticket devop40 that is dedicated to this type of discussion :)

File: fabfile/component/kraken.py [hunk 3]
```diff
@@ -253,22 +251,25 @@ def restart_kraken(instance, test=True, wait=True):
...
-        kraken = 'kraken_' + instance.name
-        start_or_stop_with_delay(kraken, 4000, 500, start=False, only_once=True)
-        start_or_stop_with_delay(kraken, 4000, 500, only_once=env.KRAKEN_START_ONLY_ONCE)
+        with settings(hosts=instance.kraken_engines):
+            kraken = 'kraken_' + instance.name
+            start_or_stop_with_delay(kraken, 4000, 500, start=False, only_once=True)
+            start_or_stop_with_delay(kraken, 4000, 500, only_once=env.KRAKEN_START_ONLY_ONCE)
         if test:
-            test_kraken(instance.name, fail_if_error=False, wait=wait)
+            test_kraken(instance, fail_if_error=False, wait=wait, hosts=instance.kraken_engines)
```
Comment (antoine-de) [inline fabfile/component/kraken.py]: I don't understand the difference between instance.kraken_engines and instance.kraken_engines_url

File: fabfile/component/kraken.py [hunk 4]
```diff
@@ -449,57 +459,72 @@ def create_eng_instance(instance):
...
+        abort("Bad 'apply_on' parameter value: {}".format(apply_on))
+
+    with settings(
+        hosts=hosts,
+        exclude_hosts=exclude_hosts,
+        warn_only=True
+    ):
+        sudo("service kraken_{} stop; sleep 3".format(instance.name))
+
+        run("update-rc.d -f kraken_{} remove".format(instance.name))
```
Comment (antoine-de) [inline fabfile/component/kraken.py]: does this work with systemd and initd ?

File: fabfile/instance.py [hunk 5]
```diff
@@ -30,38 +30,42 @@
...
+            self.kraken_zmq_socket = 'ipc://{kraken_dir}/{instance}/kraken.sock'.format(
+                kraken_dir=env.kraken_basedir, instance=self.name)
             self.jormungandr_zmq_socket_for_instance = self.kraken_zmq_socket
-        else:
-            if zmq_socket_port is not None and env.zmq_server is not None:
-                # kraken needs to listen to all but jomgandr needs to communicates with the engine
-                self.kraken_zmq_socket = 'tcp://*:{port}'.format(port=zmq_socket_port)
-                self.jormungandr_zmq_socket_for_instance = 'tcp://{server}:{port}'.format(server=env.zmq_server, port=zmq_socket_port)
-                all_zmq_ports.append(zmq_socket_port)
+            self.zmq_server, self.kraken_engines = None, env.make_ssh_url(env.roledefs['eng'])
```
Comment (antoine-de) [inline fabfile/instance.py]: I think it's a bit clearer to initialize those value in separate lines

File: tests/tests_instance.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
+# encoding: utf-8
```
Comment (antoine-de) [inline tests/tests_instance.py]: :+1: !

File: tests/tests_instance.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+import unittest
```
Comment (antoine-de) [inline tests/tests_instance.py]: I personally think py.test is better for unittest

Comment (francois-vincent) [inline tests/tests_instance.py]: Nope. unittest is defacto standard and I will use it

File: fabfile/component/kraken.py (hunk 4 above)
Comment (francois-vincent) [inline fabfile/component/kraken.py]: probably not !

File: tests/tests_instance.py (hunk 7 above)
Comment (antoine-de) [inline tests/tests_instance.py]: I don't agree with it. I find `py.test` way more powerful, flexible and clear and it's our python unit test standard :smile:

File: fabfile/component/kraken.py (hunk 3 above)
Comment (francois-vincent) [inline fabfile/component/kraken.py]: 1st is user@url 2nd is only url

Comment (antoine-de) [inline fabfile/component/kraken.py]: I don't understand why you use both in test_kraken then.

Comment (francois-vincent) [inline fabfile/component/kraken.py]: right, fixed

**Comment by Hoshiyo** [comment]
Jenkins, test this please

**Comment by Hoshiyo** [comment]
Jenkins, test this please

**Comment by antoine-de** [comment]
@francois-vincent can you explain a bit the mechanism ?

**Comment by francois-vincent** [comment]
Testé sur 2 docker !

**Comment by antoine-de** [comment]
@francois-vincent can you explain a bit the mechanism ?

**Comment by Hoshiyo** [comment]
Jenkins, test this please

**Comment by buildbotctp** [comment]
Build finished. No test results found.

**Pull request merged**